cfgmgr/buffermgrdyn.cpp

```diff
@@ -0,0 +1,1655 @@
+            case task_process_status::task_failed:
+                // We shouldn't return directly here. Because doing so will cause the following items lost
+                failed_item_count++;
+                break;
+            default:
+                break;
+            }
+
+            SWSS_LOG_DEBUG("Port Info for %s after handling speed %s cable %s gb %s",
+                           port.c_str(),
+                           portInfo.speed.c_str(), portInfo.cable_length.c_str(), portInfo.gearbox_model.c_str());
+        }
+    }
+
+    if (failed_item_count > 0)
+    {
+        return task_process_status::task_failed;
+    }
+
+    return task_process_status::task_success;
+}
+
+// A tiny state machine is required for handling the events
+// flags:
+//      speed_updated
+//      mtu_updated
+//      admin_status_updated
+// flow:
+// 1. handle all events in order, record new value if necessary
+// 2. if cable length or speed hasn't been configured, return success
+// 3. if mtu isn't configured, take the default value
+// 4. if speed_updated or mtu_updated, update headroom size
+//    elif admin_status_updated, update buffer pool size
+task_process_status BufferMgrDynamic::handlePortTable(Consumer &consumer)
+{
+    auto it = consumer.m_toSync.begin();
+    KeyOpFieldsValuesTuple t = it->second;
+    auto &port = kfvKey(t);
+    string op = kfvOp(t);
+    bool speed_updated = false, mtu_updated = false, admin_status_updated = false;
+
+    SWSS_LOG_DEBUG("processing command:%s PORT table key %s", op.c_str(), port.c_str());
+
+    port_info_t &portInfo = m_portInfoLookup[port];
+
+    SWSS_LOG_DEBUG("Port Info for %s before handling %s %s %s",
+                   port.c_str(),
+                   portInfo.speed.c_str(), portInfo.cable_length.c_str(), portInfo.gearbox_model.c_str());
+
+    task_process_status task_status = task_process_status::task_success;
+
+    if (op == SET_COMMAND)
+    {
+        for (auto i : kfvFieldsValues(t))
+        {
+            if (fvField(i) == "speed")
+            {
+                speed_updated = true;
+                portInfo.speed = fvValue(i);
+            }
+            else if (fvField(i) == "mtu")
+            {
+                mtu_updated = true;
+                portInfo.mtu = fvValue(i);
+            }
+            else if (fvField(i) == "admin_status")
+            {
+                admin_status_updated = true;
+            }
+        }
+
+        string &cable_length = portInfo.cable_length;
+        string &mtu = portInfo.mtu;
+        string &speed = portInfo.speed;
+
+        if (cable_length.empty() || speed.empty())
+        {
+            SWSS_LOG_WARN("Cable length for %s hasn't been configured yet, unable to calculate headroom", port.c_str());
+            // We don't retry here because it doesn't make sense until the cable length is configured.
+            return task_process_status::task_success;
+        }
+
+        if (speed_updated || mtu_updated)
```

Comment:
I believe we should call refreshPGs for cable length update even if speed or MTU is not updated.

Reply:
Not really.
This function is dedicated to handling the speed or MTU update.
Cable length update is handled in another function, handleCableLenTable.